### PR TITLE
Include the eth2 deposit contract as an external test

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+// supportsInterface(bytes4): 0x8564090700000000000000000000000000000000000000000000000000000000 -> true # the deposit interface id #
+// get_deposit_root() -> 0xd70a234731285c6804c2a4f56711ddb8c82c99740f207854891028af34e27e5e
+// get_deposit_count() -> 0x20, 8, 0
+// # TODO: check balance and logs after each deposit #
+// deposit(bytes,bytes,bytes,bytes32), 32 ether: 0 -> FAILURE # Empty input #
+// get_deposit_root() -> 0xd70a234731285c6804c2a4f56711ddb8c82c99740f207854891028af34e27e5e
+// get_deposit_count() -> 0x20, 8, 0
+// deposit(bytes,bytes,bytes,bytes32), 1 ether: 0x80, 0xe0, 0x120, 0xaa4a8d0b7d9077248630f1a4701ae9764e42271d7f22b7838778411857fd349e, 0x30, 0x933ad9491b62059dd065b560d256d8957a8c402cc6e8d8ee7290ae11e8f73292, 0x67a8811c397529dac52ae1342ba58c9500000000000000000000000000000000, 0x20, 0x00f50428677c60f997aadeab24aabf7fceaef491c96a52b463ae91f95611cf71, 0x60, 0xa29d01cc8c6296a8150e515b5995390ef841dc18948aa3e79be6d7c1851b4cbb, 0x5d6ff49fa70b9c782399506a22a85193151b9b691245cebafd2063012443c132, 0x4b6c36debaedefb7b2d71b0503ffdc00150aaffd42e63358238ec888901738b8 -> # txhash: 0x7085c586686d666e8bb6e9477a0f0b09565b2060a11f1c4209d3a52295033832 #
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:191]: These are real world values from https://etherscan.io/tx/0x7085c586686d666e8bb6e9477a0f0b09565b2060a11f1c4209d3a52295033832

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+// get_deposit_root() -> 0x2089653123d9c721215120b6db6738ba273bbc5228ac093b1f983badcdc8a438
+// get_deposit_count() -> 0x20, 8, 452312848583266388373324160190187140051835877600158453279131187530910662656
+// deposit(bytes,bytes,bytes,bytes32), 32 ether: 0x80, 0xe0, 0x120, 0xdbd986dc85ceb382708cf90a3500f500f0a393c5ece76963ac3ed72eccd2c301, 0x30, 0xb2ce0f79f90e7b3a113ca5783c65756f96c4b4673c2b5c1eb4efc22280259441, 0x06d601211e8866dc5b50dc48a244dd7c00000000000000000000000000000000, 0x20, 0x00344b6c73f71b11c56aba0d01b7d8ad83559f209d0a4101a515f6ad54c89771, 0x60, 0x945caaf82d18e78c033927d51f452ebcd76524497b91d7a11219cb3db6a1d369, 0x7595fc095ce489e46b2ef129591f2f6d079be4faaf345a02c5eb133c072e7c56, 0x0c6c3617eee66b4b878165c502357d49485326bc6b31bc96873f308c8f19c09d -> # txhash: 0x404d8e109822ce448e68f45216c12cb051b784d068fbe98317ab8e50c58304ac #
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:194]: These are real world values from https://etherscan.io/tx/0x404d8e109822ce448e68f45216c12cb051b784d068fbe98317ab8e50c58304ac

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+// compileViaYul: also
+// ----
+// constructor()
+// supportsInterface(bytes4): 0x0 -> 0
+// supportsInterface(bytes4): 0xffffffff00000000000000000000000000000000000000000000000000000000 -> false # defined to be false by ERC-165 #
+// supportsInterface(bytes4): 0x01ffc9a700000000000000000000000000000000000000000000000000000000 -> true # ERC-165 id #
+// supportsInterface(bytes4): 0x8564090700000000000000000000000000000000000000000000000000000000 -> true # the deposit interface id #
+// get_deposit_root() -> 0xd70a234731285c6804c2a4f56711ddb8c82c99740f207854891028af34e27e5e
+// get_deposit_count() -> 0x20, 8, 0
+// # TODO: check balance and logs after each deposit #
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:187]: #10426 and #10427 would greatly augment this test

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        ret[2] = bytesValue[5];
+        ret[3] = bytesValue[4];
+        ret[4] = bytesValue[3];
+        ret[5] = bytesValue[2];
+        ret[6] = bytesValue[1];
+        ret[7] = bytesValue[0];
+    }
+}
+// ====
+// compileViaYul: also
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:178]: With IR enabled:
```
real	0m7.803s
```

Without IR:
```
real	0m0.584s
```

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+// ----
+// constructor()
+// supportsInterface(bytes4): 0x0 -> 0
+// supportsInterface(bytes4): 0xffffffff00000000000000000000000000000000000000000000000000000000 -> false # defined to be false by ERC-165 #
+// supportsInterface(bytes4): 0x01ffc9a700000000000000000000000000000000000000000000000000000000 -> true # ERC-165 id #
+// supportsInterface(bytes4): 0x8564090700000000000000000000000000000000000000000000000000000000 -> true # the deposit interface id #
+// get_deposit_root() -> 0xd70a234731285c6804c2a4f56711ddb8c82c99740f207854891028af34e27e5e
+// get_deposit_count() -> 0x20, 8, 0
+// # TODO: check balance and logs after each deposit #
+// deposit(bytes,bytes,bytes,bytes32), 32 ether: 0 -> FAILURE # Empty input #
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:188]: Actually the test parser doesn't allow empty input. This fails parsing:
```
// deposit(bytes,bytes,bytes,bytes32), 32 ether: -> FAILURE # Empty input #
```

With:
```
semanticTests/externalContracts/deposit_contract.sol: Exception during test: Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: solidity::frontend::test::TestParserError
std::exception::what: Line 188: No argument provided.
[solidity::util::tag_comment*] = Line 188: No argument provided.
```

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol [hunk 6]
```diff
@@ -0,0 +1,196 @@
+// ┏━━━┓━┏┓━┏┓━━┏━━━┓━━┏━━━┓━━━━┏━━━┓━━━━━━━━━━━━━━━━━━━┏┓━━━━━┏━━━┓━━━━━━━━━┏┓━━━━━━━━━━━━━━┏┓━
+// ┃┏━━┛┏┛┗┓┃┃━━┃┏━┓┃━━┃┏━┓┃━━━━┗┓┏┓┃━━━━━━━━━━━━━━━━━━┏┛┗┓━━━━┃┏━┓┃━━━━━━━━┏┛┗┓━━━━━━━━━━━━┏┛┗┓
+// ┃┗━━┓┗┓┏┛┃┗━┓┗┛┏┛┃━━┃┃━┃┃━━━━━┃┃┃┃┏━━┓┏━━┓┏━━┓┏━━┓┏┓┗┓┏┛━━━━┃┃━┗┛┏━━┓┏━┓━┗┓┏┛┏━┓┏━━┓━┏━━┓┗┓┏┛
+// ┃┏━━┛━┃┃━┃┏┓┃┏━┛┏┛━━┃┃━┃┃━━━━━┃┃┃┃┃┏┓┃┃┏┓┃┃┏┓┃┃━━┫┣┫━┃┃━━━━━┃┃━┏┓┃┏┓┃┃┏┓┓━┃┃━┃┏┛┗━┓┃━┃┏━┛━┃┃━
+// ┃┗━━┓━┃┗┓┃┃┃┃┃┃┗━┓┏┓┃┗━┛┃━━━━┏┛┗┛┃┃┃━┫┃┗┛┃┃┗┛┃┣━━┃┃┃━┃┗┓━━━━┃┗━┛┃┃┗┛┃┃┃┃┃━┃┗┓┃┃━┃┗┛┗┓┃┗━┓━┃┗┓
+// ┗━━━┛━┗━┛┗┛┗┛┗━━━┛┗┛┗━━━┛━━━━┗━━━┛┗━━┛┃┏━┛┗━━┛┗━━┛┗┛━┗━┛━━━━┗━━━┛┗━━┛┗┛┗┛━┗━┛┗┛━┗━━━┛┗━━┛━┗━┛
+// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┃┃━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┗┛━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+// SPDX-License-Identifier: CC0-1.0
```
Comment (axic) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:10]: The test failure is because the test suite adds an SPDX line, so need to remove this, or fix the test suite to not add one if it is present.

File: test/libsolidity/AnalysisFramework.cpp [hunk 7]
```diff
@@ -51,9 +51,11 @@ AnalysisFramework::parseAnalyseAndReturnError(
 )
 {
 	compiler().reset();
+	// Do not insert license if it is already present.
+	bool insertLicense = _insertLicenseAndVersionPragma && _source.find("// SPDX-License-Identifier:") == string::npos;
```
Comment (axic) [inline test/libsolidity/AnalysisFramework.cpp:55]: Actually we should have a generic "prepareSources" helper for all the different tests. Will do that in a separate PR.

**Comment by chriseth** [comment]
Maybe target `breaking` already?

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 8]
```diff
@@ -132,7 +132,9 @@ bytes SolidityExecutionFramework::compileContract(
 string SolidityExecutionFramework::addPreamble(string const& _sourceCode)
 {
 	// Silence compiler version warning
-	string preamble = "pragma solidity >=0.0;\n// SPDX-License-Identifier: unlicensed\n";
+	string preamble = "pragma solidity >=0.0;\n";
+	if (_sourceCode.find("// SPDX-License-Identifier:") == string::npos)
+		preamble += "// SPDX-License-Identifier: unlicensed\n";
```
Comment (axic) [inline test/libsolidity/SolidityExecutionFramework.cpp:137]: Interesting, we use `unlicensed` here, but `GPL-3.0` in `AnalysisFramework`.

**Comment by axic** [comment]
> Maybe target breaking already?

I think it is fine to develop, it would at least show us if there's any difference between the two?

**Comment by chriseth** [comment]
@axic please at least check that it compiles with 0.8.x

**Comment by axic** [comment]
> please at least check that it compiles with 0.8.x

Trying it now.

**Comment by axic** [comment]
It seems to be passing.

**Review by chriseth — APPROVED** [review]

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 8 above)
Comment (cameel) [reply test/libsolidity/SolidityExecutionFramework.cpp:137]: Should we change that to GPL3?

Comment (axic) [reply test/libsolidity/SolidityExecutionFramework.cpp:137]: I think so, but it does not matter too much besides consistency.
